Cleanup conversion from ASCII-HEX to binary

dcooper16 opened the pull request:
This PR makes some minor improvements to code that converts ASCII-HEX to binary.

First, testssl.sh currently has two functions that do the same thing: `asciihex_to_binary()` and `hex2ascii()`. This PR converts all calls to either of these functions with calls to `hex2binary()`, which is based on the current `asciihex_to_binary()`.

This commit also changes direct use of `printf` to calls to `hex2binary()` in `generate-ccm-counter-blocks()`, `ccm-compute-tag()`, and `generate_gcm_counter()`.

Finally, this commit cleans up the implementation of `hex2binary()` a bit and introduces the option to use `xxd`, if that program is available. Using `xxd` rather than multiple calls to `printf` has a couple of advantages. `xxd` is a bit faster. However, the primary advantage is that when debugging (i.e., using `bash -x`), each call to `hex2binary()` only includes a few steps, regardless of how long the string is, whereas using `printf` the call to `hex2binary()` could take hundreds of steps.

Commit:
Cleanup conversion from ASCII-HEX to binary

This commit makes some minor improvements to code that converts ASCII-HEX to binary.

First, testssl.sh currently has two functions that do the same thing: asciihex_to_binary() and hex2ascii(). This commit all calls to either of these functions with calls to hex2binary(), which is based on the current asciihex_to_binary().

This commit also changes direct use of printf to calls to hex2ascii() in generate-ccm-counter-blocks(), ccm-compute-tag(), and generate_gcm_counter().

Finally, this commit cleans up the implemention of hex2binary() a bit and introduces the option to use xxd, if that program is available. Using xxd rather than multiple calls to printf has a couple of advantages. xxd is a bit faster. However, the primary advantage is that when debugging (i.e., using bash -x), each call to hex2binary() only includes a few steps, regardless of how long the string is, whereas using printf the call to hex2binary() could take hundreds of steps.

drwetter commented:
Hi David,

good idea,thanks!  One remark though: why ``hex2binary()``? I find the binary part a bit misleading.

(for Travis failure, see #1855 . Don't mind because of this PR)

Cheers, Dirk

dcooper16 commented:
> why `hex2binary()`?

In most cases the function is being called with data such as certificates, CRLs, OCSP responses, public keys, or ciphertext. There are only a few places, such as where `hex2binary()` is replacing `hex2ascii()` in which the data being converted represents text (i.e., ASCII).

Do you have any suggestions for a different name?

drwetter commented:
Good point. :)  Maybe ``hex2x`` adding a comment that the trailing x can be either ASCII or binary?

Commit:
Reintroduce hex2ascii()

Add hex2ascii() as a function that just calls hex2binary().

dcooper16 commented:
Hi Dirk,

I don't think that `hex2x()` would be any less confusing, as people will probably try to guess what that means before going to the function definition and reading the comment.

How about just having two functions, `hex2binary()` and `hex2ascii()`. I've updated the PR so that the previous calls to `hex2ascii()` remain unchanged, but `hex2ascii()` is now just defined as a call to `hex2binary()`. I also added more text to the comment at the beginning of `hex2binary()`.

drwetter commented:
Sounds reasonable, thanks!

One last thing maybe: Your xxd shortcut sounded like an excellent idea to me! Do you mind to (re)add that to ``hex2ascii?()`` ? The function is not much in use outside debug statements. But if the usage changes ... and for the sake providing good code.

I can take that on my list as well but I'd need a bit

dcooper16 commented:
> One last thing maybe: Your xxd shortcut sounded like an excellent idea to me! Do you mind to (re)add that to `hex2ascii?()`? 

Hi Dirk,

In the PR `hex2ascii()` is just defined as:
```
# convert 414243 into ABC
hex2ascii() {
     hex2binary $1
}
```
This adds the overhead of an extra function call, but I figured it was okay since `hex2ascii()` is rarely used.

drwetter commented:
Thanks for pointing me to the correct spot  ;-/
